meta-security-framework/classes/xattr-images.bbclass

```diff
@@ -15,3 +23,89 @@ EXTRA_IMAGECMD_jffs2_append = " --with-xattr"
 IMAGE_DEPENDS_tar_append = " tar-replacement-native"
 EXTRANATIVEPATH += "tar-native"
 IMAGE_CMD_TAR = "tar --xattrs --xattrs-include=*"
+
+xattr_images_fix_transmute () {
+    set -e
+    cd ${IMAGE_ROOTFS}
+
+    # The recursive updating of the Smack label ensures that each entry
+    # has the label set for its parent directories if one of those was
+    # marked as transmuting.
+    #
+    # In addition, "-" is set explicitly on everything that would not
+    # have a label otherwise. This is a workaround for tools like swupd
+    # which transfers files from a rootfs onto a target device where Smack
+    # is active: on the target, each file gets assigned a label, typically
+    # the one from the process which creates it. swupd (or rather, the tools
+    # it is currently built on) knows how to set security.SMACK64="_" when
+    # it is set on the original files, but it does not know that it needs
+    # to remove that xattr when not set.
+    python <<EOF
+import os
+import errno
+
+# Cannot use the 'xattr' module, it is not part of a standard Python
+# installation. Instead re-implement using ctypes. Only has to be good
+# enough for xattrs that are strings. Always operates on the symlinks themselves,
+# not what they point to.
+import ctypes
+
+# We cannot look up the xattr functions inside libc. That bypasses
+# pseudo, which overrides these functions via LD_PRELOAD. Instead we have to
+# find the function address and then create a ctypes function from it.
+libdl = ctypes.CDLL("libdl.so.2")
+_dlsym = libdl.dlsym
+_dlsym.restype = ctypes.c_void_p
+RTLD_DEFAULT = ctypes.c_void_p(0)
+_lgetxattr = ctypes.CFUNCTYPE(ctypes.c_ssize_t, ctypes.c_char_p, ctypes.c_char_p, ctypes.c_void_p, ctypes.c_size_t,
+            use_errno=True)(_dlsym(RTLD_DEFAULT, 'lgetxattr'))
+_lsetxattr = ctypes.CFUNCTYPE(ctypes.c_int, ctypes.c_char_p, ctypes.c_char_p, ctypes.c_void_p, ctypes.c_size_t, ctypes.c_int,
+            use_errno=True)(_dlsym(RTLD_DEFAULT, 'lsetxattr'))
+
+def lgetxattr(f, attr, default=None):
+    len = 32
```

Comment:
Because looping isn't hard to implement and I did not want to make lgetxattr() depend on Smack constraints. As it stands now, it is a general-purpose method.